### DISCUSSION TITLE
Freezes Windows 11 when open a pengwin terminal

### DISCUSSION THREAD

**Discussion by sn4k3** [WhitewaterFoundry/Pengwin#726] [Q&A] [answered]
I've made a fresh install of Windows 11 on two computers, a laptop and a desktop.
When i run the pengwin app it slows down all windows tasks by a ton of time, for example, open the taskmanager takes about 5m, as well other programs.
However pengwin terminal remains fast.
If i close pengwin after 1 minute windows unstuck and start working normally.

Anyone else have this problem?
It was great on previous Windows 10 install.

Both was fresh installs

**Comment by crramirez** [comment]
Hello @sn4k3,

Please tell me if pengwin is in WSL 1 or 2:

`wsl.exe --list --verbose`

There is a known problem about memory in WSL 1 with Windows 11

**Reply by sn4k3** [reply]
I made fresh installs on Windows 11, it comes with WSL 2 by default.
Got from the official from MS Store

```
C:\Users\tiago>wsl.exe --list --verbose
  NAME      STATE           VERSION
* WLinux    Stopped         2
```

Note: If i run a command like: `wsl zip something to.zip` command runs without freeze windows. Problem is when i open pengwin terminal

**Reply by crramirez** [reply]
Can you please leave a terminal with the command `top` executing and open another? Can you see any spike?

I haven't seen this issue before, but someone left a review with the same issue.

**Reply by sn4k3** [reply]
See how bad it is: https://youtu.be/GweRvtm0eM0
It only affect new spawned processes, the already running processes are fine to work with. Both CPU, ram and disk kept at low usage, no spikes while lunching Pengwin terminal. All CPU load you can see is in fact the screen recorder.
Note that it only return to normal when the second task manager loads and show the processes.

On my laptop i installed pengwin on a clean install without any OS mess or drivers and i had the exact same problem

**Reply by crramirez** [reply]
Ok, I was trying to reproduce the error without success. The first step would be disabling the Pengwin startup script.
Please edit with sudo the following file `/etc/profile.d/00-pengwin.sh` and add **return** as the second line after `#!/bin/sh`

```bash
#!/bin/sh

return

# Only the default WSL user should run this script
if ! (id -Gn | grep -c "adm.*sudo\|sudo.*adm" >/dev/null); then
  return
fi

setup_display() {

  if [ -n "${XRDP_SESSION}" ]; then
....
```


Restart Pengwin and see if it improves.

Have you tested with other distros?

What about `wsl --update` have you tried it?

Regards,
Carlos

**Reply by sn4k3** [reply]
> Have you tested with other distros?

On laptop before install pengwin i installed Ubuntu as mistake and i run it to test, it does not raise the freeze problem.
Can it be realated to some of the Windows explorer extensions that are part of pengwin synergy?

> What about wsl --update have you tried it?

Just did, but no updates...

```
PS C:\Users\tiago> wsl --update
Checking for updates...
No updates are available.
Kernel version: 5.10.60.1
```

> Please edit with sudo the following file /etc/profile.d/00-pengwin.sh and add return as the second line after #!/bin/sh
Restart Pengwin and see if it improves.

Yup... Completly solve the problem! No more freezing the processes running that way :)

**Reply by crramirez** [reply]
Ok, could you please run this: `bash -x /etc/profile.d/00-pengwin.sh` and send me the result? You can send it to contact@whitewaterfoundry.com so your data is not pasted here. Please notice while the script is executing if you see some delays. Also, if after running the script the freezes return.

Remember to comment out the second line with the **return** so the script is executed.

Regards,
Carlos

**Reply by sn4k3** [reply]
Didnt see any critical data... Heres the output:

```
tiago@TIAGO-PC:~$ bash -x /etc/profile.d/00-pengwin.sh
+ id -Gn
+ grep -c 'adm.*sudo\|sudo.*adm'
+ setup_display
+ '[' -n '' ']'
+ '[' -n /run/WSL/8_interop ']'
+ '[' -n :0 ']'
+ sudo /usr/local/bin/wsl_change_checker 3 WSL2 :0
++ echo :0
++ cut -d : -f 1
+ sudo /usr/local/bin/wsl2_ip_checker ''
+ export WSL2=3
+ WSL2=3
+ return
+ command -v glxinfo
+ unset LIBGL_ALWAYS_INDIRECT
+ sudo /usr/local/bin/libgl-change-checker 0
+ export NO_AT_BRIDGE=1
+ NO_AT_BRIDGE=1
+ alias 'clear=clear -x'
+ alias 'll=ls -al'
+ alias 'winget=powershell.exe winget'
+ alias wsl=wsl.exe
+ command -v cmd.exe
+ '[' '!' -f /home/tiago/.firstrun ']'
+ wslpath 'C:\'
++ cmd.exe /c 'cd %SYSTEMDRIVE%\ && echo %HOMEDRIVE%%HOMEPATH%'
++ tr -d '\r'
+ wHomeWinPath='C:\Users\tiago'
+ '[' 14 -le 3 ']'
++ wslpath -u 'C:\Users\tiago'
+ export WIN_HOME=/mnt/c/Users/tiago
+ WIN_HOME=/mnt/c/Users/tiago
+ win_home_lnk=/home/tiago/winhome
+ '[' '!' -e /home/tiago/winhome ']'
+ unset win_home_lnk
```

A delay happens after line `++ tr -d '\r'` about 5s, everything else shows fast. After execute that the freeze problem on windows happens

**Comment by crramirez** [comment]
Hello @sn4k3,

Thanks for testing. 5s in fact is too much time, this is the only call to a Windows process in the script. I have seen interop problems but with the WSL Store edition Preview.

Could you then make som tests executing `cmd.exe /c 'cd %SYSTEMDRIVE%\ && echo %HOMEDRIVE%%HOMEPATH%' 2>/dev/null | tr -d '\r'` in Ubuntu for example to see if the slow down happens. Also, you can add again the return statement at the beginning of the script and run `cmd.exe /c 'cd %SYSTEMDRIVE%\ && echo %HOMEDRIVE%%HOMEPATH%' 2>/dev/null | tr -d '\r'` in the command prompt to see what happens.

Regards,
Carlos

**Reply by sn4k3** [reply]
Running the command on pengwin raises the problem:

```
tiago@TIAGO-PC:~$ time cmd.exe /c 'cd %SYSTEMDRIVE%\ && echo %HOMEDRIVE%%HOMEPATH%' 2>/dev/null | tr -d '\r'
C:\Users\tiago

real    0m4.586s
user    0m0.010s
sys     0m0.000s
```

I made simpler test, and run `cmd.exe /c 'echo lol'` does also raise the problem :(
Will try on ubuntu latter

**Reply by sn4k3** [reply]
Update: On ubuntu the same command raises the same problem and delay :(

**Reply by crramirez** [reply]
Ok, good we have a triangle now. Can you please now test the following command? Only works on Pengwin:

`time cmd-exe /c 'cd %SYSTEMDRIVE%\ && echo %HOMEDRIVE%%HOMEPATH%' 2>/dev/null | tr -d '\r'`
`#-------^ notice the dash`

If it doesn't raise the problem I can release a hotfix.

This issue has almost the same behavior as this: https://github.com/microsoft/WSL/issues/7883 and https://github.com/microsoft/WSL/issues/7754 but those are related to the WSL Store version, it happens to me as well.

Regards,
Carlos

**Reply by sn4k3** [reply]
It works without the problem :)

```
tiago@TIAGO-PC:~$ time cmd-exe /c 'cd %SYSTEMDRIVE%\ && echo %HOMEDRIVE%%HOMEPATH%' 2>/dev/null | tr -d '\r'
C:\Users\tiago

real    0m0.088s
user    0m0.005s
sys     0m0.007s
```

I installed pengwin via Store (https://www.microsoft.com/en-us/p/pengwin/9nv1gv1pxz6p?activetab=pivot:overviewtab), but WSL was installed and prepared via terminal.

**Reply by crramirez** [reply]
Thanks a lot for your support. This weekend I'll release a new patch version with the change. You can see it applied in the development branch.

Now you deserve an explanation. The problem is on the service that allows you explorer the Linux file system from Windows Explorer. When you execute cmd.exe from the Linux home it complains about the network path like \\wsl.localhost. In this instant the connection between Windows filesystem and the Linux one is established.

cmd-exe what it does is to first change the current directory to C:\ and then execute the command. This avoid that cmd be executed in the network path. cmd-exe was originally created to avoid the error message, but it has resulted more useful now.

The weird part is that I cannot reproduce it. 

Regards,
Carlos

**Reply by sn4k3** [reply]
Thank you for the assistance and troubleshoot! I'm glad to help and end in a solution 👍
It was a big pain every time i had to open that terminal :/
For now i will edit the script to use the `cmd-exe`

Best regards

**Comment by crramirez** [comment, marked as answer]
Run `pengwin-setup update` to get the updated scripts. Remember telling "Package version" when it asks for `00-pengwin.sh` replacement.

**Reply by sn4k3** [reply]
It seens one line was not changed, maybe not required... It runs fine now but:

```
--- /etc/profile.d/00-pengwin.sh        2022-01-28 20:23:25.130752800 +0000
+++ /etc/profile.d/00-pengwin.sh.dpkg-new       2022-01-28 22:39:09.000000000 +0000
@@ -98,7 +98,7 @@
 alias wsl='wsl.exe'

 # Check if we have Windows Path
-if (command -v cmd-exe >/dev/null); then
+if (command -v cmd.exe >/dev/null); then

   # Execute on user's shell first-run
   if [ ! -f "${HOME}/.firstrun" ]; then

Configuration file '/etc/profile.d/00-pengwin.sh'
```

**Reply by crramirez** [reply]
You can directly replace the script from the new one with:

`sudo rm /etc/profile.d/00-pengwin.sh && sudo curl https://raw.githubusercontent.com/WhitewaterFoundry/pengwin-base/master/profile.d/00-pengwin.sh -o /etc/profile.d/00-pengwin.sh`

**Comment by CaptainFrosty** [comment]
Just throwing it out there, but I had issues with Pengwin crashes and Windows 11 and they seemed to go away when I uninstalled the built-in WSL and only used the Store preview version. Microsoft says having them both is no problem at all, but they say a lot of things. My system did not seem to like having both. Just a suggestion if it applies to you.

**Reply by crramirez** [reply]
Interesting. I have experienced a lot of freezes with the Store version especially when running intellij's auto jdk detection. But I haven't tested uninstalling the built in one. I know that if I uninstall the store version the problem goes away.

It worth testing and maybe the problem is she you have both versions?
